hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/replication/VerifyReplication.java

```diff
@@ -240,55 +266,138 @@ public void map(ImmutableBytesWritable row, final Result value, Context context)
+      }
+
+      @Override
+      public void run() {
+        Get get = new Get(row);
+        get.setCacheBlocks(tableScan.getCacheBlocks());
+        get.setFilter(tableScan.getFilter());
+
+        int sleepMs = sleepMsBeforeReCompare;
+        int tries = 0;
+
+        while (++tries <= reCompareTries) {
+          context.getCounter(Counters.RE_COMPARES).increment(1);
+
+          try {
+            Thread.sleep(sleepMs);
+          } catch (InterruptedException e) {
+            LOG.warn("Sleeping interrupted, incrementing bad rows and aborting");
+            incrementOriginalAndBadCounter();
+            Thread.currentThread().interrupt();
+          }
+
+          try {
+            fetchLatestRows(get);
+            if (matches(sourceResult, replicatedResult, null)) {
+              if (verbose) {
+                LOG.info("Good row key (with recompare): " + delimiter + Bytes.toStringBinary(row)
+                  + delimiter);
+              }
+              context.getCounter(Counters.GOODROWS).increment(1);
+              return;
+            }
+          } catch (IOException e) {
+            context.getCounter(Counters.FAILED_RECOMPARE).increment(1);
             if (verbose) {
-              LOG.info("Good row key (with recompare): " + delimiter
-                + Bytes.toStringBinary(row.getRow()) + delimiter);
+              LOG.info("Got an exception during recompare for rowkey=" + Bytes.toStringBinary(row),
+                e);
             }
           }
-          return;
+
+          sleepMs = sleepMs * (2 ^ reCompareBackoffExponent);
+        }
+
+        LOG.error(originalCounter.toString() + ", rowkey=" + delimiter + Bytes.toStringBinary(row) + delimiter);
+        incrementOriginalAndBadCounter();
+      }
+
+      private void fetchLatestRows(Get get) throws IOException {
+        int tries = 0;
+        while (++tries <= FETCH_LATEST_ROWS_TRIES) {
+          Result sourceResult = sourceTable.get(get);
+          Result replicatedResult = replicatedTable.get(get);
+
+          if (matches(sourceResult, this.sourceResult, Counters.SOURCE_ROW_CHANGED) &&
+            matches(replicatedResult, this.replicatedResult, Counters.PEER_ROW_CHANGED)) {
+            return;
+          }
+
+          this.sourceResult = sourceResult;
+          this.replicatedResult = replicatedResult;
+          Threads.sleep(FETCH_LATEST_ROWS_BACKOFF_MILLIS);
+        }
+      }
+
+      private boolean matches(Result original, Result updated, Counters failCounter) {
+        try {
+          Result.compareResults(original, updated);
+          return true;
         } catch (Exception e) {
-          LOG.error("recompare fail after sleep, rowkey=" + delimiter
-            + Bytes.toStringBinary(row.getRow()) + delimiter);
+          if (failCounter != null) {
+            context.getCounter(failCounter).increment(1);
+            LOG.debug(failCounter + " for rowkey=" + Bytes.toStringBinary(row));
```

Comment:
wrap this in LOG.isDebugEnabled, since converting bytes to StringBinary takes cpu

Reply:
I opted for checking `if (verbose)` just for consistency's sake, let me know if that's alright.